src/foam/dao/PipelinePMDAO.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+import foam.nanos.pm.PM;
+
+public class PipelinePMDAO
+  extends ProxyDAO
+{
+  protected String putName_;
+  protected String findName_;
+  protected String removeName_;
+  protected String removeAllName_;
+  protected String delegateName_;
+
+  public PipelinePMDAO(X x, DAO delegate) {
+    super(x, delegate);
+    delegateName_ = getDelegate().getClass().getName();
+    init();
+  }
+
+  /** Calls a method to create the PM pipeline and sets the name variables of each operation of interest */
+  void init() {
+    createPipeline();
+    putName_       = delegateName_ + ":pipePut";
+    findName_      = delegateName_ + ":pipeFind";
+    removeName_    = delegateName_ + ":pipeRemove";
+    removeAllName_ = delegateName_ + ":pipeRemoveAll";
+  }
+
+  /** Creates the PM pipeline by adding an EndPipelinePMDAO after of this class only if it is a ProxyDAO. 
+  *   If the delegate of that is also a ProxyDAO, creates a new PipelinedPMDAO in the chain beofre it which repeats this procedure recursively. */
+  private void createPipeline() {
+    DAO delegate = getDelegate();
+    DAO secondaryDelegate;
+    if ( delegate instanceof ProxyDAO ) {
```

Comment:
Put this check in EasyDAO instead. PMPipelineDAO shouldn't be called on non-decorators.